R/mol_roi.R

```diff
@@ -0,0 +1,253 @@
+#'
+#' @importFrom sf st_bbox st_as_sf st_as_sfc st_within
+#' @importFrom stringr str_detect
+#' @importFrom terra crs rast geom vect
+#'
+#' @export mol_roi
+#'
+#' @examples
+#' x <- data.frame(
+#'   decimalLatitude = c(1, 2, 3, 5),
+#'   decimalLongitude = c(1, 2, 3, 4)
+#' )
+#'
+#' roi <- sf::st_as_sfc(sf::st_bbox(c(xmin = 0, xmax = 4, ymax = 4, ymin = 0), crs = 4326))
+#' 
+#' mol_roi(
+#'   data = x,
+#'   roi  = roi,
+#'   lat = "decimalLatitude",
+#'   lon = "decimalLongitude"
+#' )
+#'
+
+mol_roi <- 
+  function(data,
+           roi,
+           lat = "decimalLatitude",
+           lon = "decimalLongitude",
+           byExtOnly = FALSE,
+           maskValue = NA){
+    
+    ### Run checks
+    if (!is.data.frame(data)) {
+      stop(deparse(substitute(data)), " is not a data.frame", call. = FALSE)
+    }
+    
+    check_col(data, c(lat, lon))
+    
+    if(!any(c("character",
+              "sf",
+              "sfc",
+              "SpatialPolygons",
+              "RasterLayer",
+              "SpatRaster")) %in% class(roi)){
+      stop("The region of interest is not in a valid format.")
+    }
+    
+    ### converts coordinates columns to numeric
+    data <-
+      data %>%
+      dplyr::rowid_to_column("id_temp") %>%
+      dplyr::mutate(decimalLatitude = as.numeric(.data[[lat]]),
+                    decimalLongitude = as.numeric(.data[[lon]]),
+                    .outside_roi = TRUE) 
+    
+    ### Screen NAs
+    dataCoords %>% 
+      dplyr::select(dplyr::all_of(c("id_temp", lon, lat))) %>%
+      dplyr::filter(!is.na(.data[[lat]]), 
+                    !is.na(.data[[lon]]))
+    
+    if(nrow(dataCoords) == 0) stop("No valid coordinates.")
+    
+    ### Read file path
+    if(is.character(roi)){
+      
+      # check file exists
+      if(!file.exists(roi)) stop("The ROI filepath is invalid.")
+      
+      # check proper format
+      if(!any(str_detect(roi,
+                     c("\\.shp$", "\\.gpkg$", "\\.tif$")))){
+        stop("ROI input files must have '.shp', '.gpkg' or '.tif' file extensions.")
+      }
+      
+      # load roi
+      if(str_detect(roi, "\\.tif")){
+        roi <- terra::rast(roi)
+      } else{
+        roi <- st_read(roi)
+      }
+      
+    }
+    
+    ### Convert to sf/terra
+    if("SpatialPolygons" %in% .class2(roi)) roi <- st_as_sfc(roi)
+    if(class(roi) == "RasterLayer") roi <- rast(roi)
+    
+    ### Run appropriate check
+    if(any(c("sf", "sfc") %in% class(roi))){
+      
+      if(!any(is.na(maskValue), is.null(maskValue))){
+        warning("maskValue is ignored when ROI is provided as a vector layer.")
+      }
+      
+      unflagged <- roi_sf(dataCoords,
+                          roi,
+                          lat,
+                          lon)
+    } else(
+      unflagged <- roi_rast(dataCoords,
+                            roi,
+                            lat,
+                            lon)
+    )
```

Comment:
There is an error here 

Spherical geometry (s2) switched off
although coordinates are longitude/latitude, st_union assumes that they are planar
Spherical geometry (s2) switched on
Error in `dplyr::mutate()`:
! Problem while computing `within = st_within(., roi, sparse = F)[, 1]`.
Caused by error in `UseMethod()`:
! no applicable method for 'st_geometry' applied to an object of class "data.frame"

Reply:
Hi @matthewsrogan,

I would like to work on some suggested functions to turn them able to bdc. May I make pull requests to the forked repo bdc_mol?

Reply:
Hi Lucas, that would be great. I just pushed some updates I've been testing this week.